### DISCUSSION TITLE
Correct cleanup to avoid memory leaks in basic TLS client doing GET/POST

### DISCUSSION THREAD

**Discussion by chr0meice2** [openssl/openssl#22305] [Q&A]
Hello All,

I am using version 3.0.11 and have a question about correct cleanup.   Using this as static library in Windows EXE and was compiled with error strings disabled.  At a very high-level, this program does nothing more than make GET/POST to a server using TLS... very basic.  Our app is currently working fine and doesn't appear leaking but want to double check no silly mistakes.
We call SSL_library_init()   at program startup and nothing else

When exiting the software, we call these functions in this order:
      WSACleanup();

   	EVP_cleanup();

    	CRYPTO_cleanup_all_ex_data();    

For the function that does the actual work in TLS GET/POST:

The only objects we use are:
   SSL_CTX* ctx = NULL;
    SSL*     ssl = NULL;

We free these at end of function and when errors might occur:

    if (ssl) { SSL_free(ssl);   }
    if (ctx) {  SSL_CTX_free(ctx);    }

We are only using these  7 functions in the library which are in a single function.
SSL_set_fd
SSL_CTX_set_verify
SSL_CTX_new
SSL_read
SSL_write
SSL_connect
SSL_get_error

Please let me know if this seems correct or if should add any addition cleanup.  Basically trying to be 100% sure no memory leaks or silly issues.  Our app appears to work fine but want to double check to be safe.

Thanks a ton!

-AT

**Comment by paulidale** [comment]
You shouldn't need to call `SSL_library_init()` at all, initialisation is done on demand.
Likewise, explicit cleanup/shutdown isn't required.  So long as you free everything that was allocated, the rest should just happen.

**Comment by mattcaswell** [comment]
> We call SSL_library_init() at program startup and nothing else

You don't need to call this. In old versions of OpenSSL this was required - but no longer. It was deprecated in OpenSSL 1.1.0.

> EVP_cleanup();
>
> CRYPTO_cleanup_all_ex_data();  

You don't need to call either of these functions. They are no-ops and deprecated since OpenSSL 1.1.0.

> if (ssl) { SSL_free(ssl);   }
> if (ctx) {  SSL_CTX_free(ctx);    }

The "if" part is unnecessary (but not harmful) - you can just call `SSL_free` and `SSL_CTX_free` directly. It is valid to call those functions with a NULL pointer if you wish.
